File: backend/app/routers/notifications.py
from fastapi import APIRouter, Depends, HTTPException

from app.ai_watch import archive_notification, mark_notification_actioned, mark_notification_read, run_ai_watch_engine
from app.auth import require_roles
from app.db import get_db

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get("")
async def get_notifications(user: dict = Depends(require_roles("admin", "manager", "member", "viewer"))) -> dict:
    with get_db() as db:
        return run_ai_watch_engine(db, int(user["id"]))


@router.post("/run-watch")
async def run_watch(user: dict = Depends(require_roles("admin", "manager", "member", "viewer"))) -> dict:
    with get_db() as db:
        return run_ai_watch_engine(db, int(user["id"]))


@router.patch("/{notification_id}/read")
async def patch_notification_read(notification_id: int, user: dict = Depends(require_roles("admin", "manager", "member", "viewer"))) -> dict:
    with get_db() as db:
        notification = mark_notification_read(db, notification_id, int(user["id"]))
    if not notification:
        raise HTTPException(status_code=404, detail="通知が見つかりません。")
    return {"notification": notification}


@router.patch("/{notification_id}/actioned")
async def patch_notification_actioned(notification_id: int, user: dict = Depends(require_roles("admin", "manager", "member", "viewer"))) -> dict:
    with get_db() as db:
        notification = mark_notification_actioned(db, notification_id, int(user["id"]))
    if not notification:
        raise HTTPException(status_code=404, detail="通知が見つかりません。")
    return {"notification": notification}


@router.patch("/{notification_id}/archive")
async def patch_notification_archive(notification_id: int, user: dict = Depends(require_roles("admin", "manager", "member", "viewer"))) -> dict:
    with get_db() as db:
        notification = archive_notification(db, notification_id, int(user["id"]))
    if not notification:
        raise HTTPException(status_code=404, detail="通知が見つかりません。")
    return {"notification": notification}

File: frontend/client-api/notifications.ts
import { fetchJson } from "@/client-api/client";
import type { AiNotification, AiNotificationCenterData } from "@/types/app";

export function getAiNotifications(): Promise<AiNotificationCenterData> {
  return fetchJson("/api/notifications");
}

export function runAiWatchEngine(): Promise<AiNotificationCenterData> {
  return fetchJson("/api/notifications/run-watch", { method: "POST" });
}

export function markAiNotificationRead(notificationId: number): Promise<{ notification: AiNotification }> {
  return fetchJson(`/api/notifications/${notificationId}/read`, { method: "PATCH" });
}

export function markAiNotificationActioned(notificationId: number): Promise<{ notification: AiNotification }> {
  return fetchJson(`/api/notifications/${notificationId}/actioned`, { method: "PATCH" });
}

export function archiveAiNotification(notificationId: number): Promise<{ notification: AiNotification }> {
  return fetchJson(`/api/notifications/${notificationId}/archive`, { method: "PATCH" });
}
